Network/HTTP2/H2/Sender.hs

```diff
@@ -200,17 +206,30 @@ frameSender
             let payloadOff = off0 + frameHeaderLength
                 datBuf = confWriteBuffer `plusPtr` payloadOff
                 datBufSiz = buflim - payloadOff
-            Next datPayloadLen reqflush mnext <- curr datBuf (min datBufSiz lim)
-            NextTrailersMaker tlrmkr' <- runTrailersMaker tlrmkr datBuf datPayloadLen
-            fillDataHeaderEnqueueNext
-                strm
-                off0
-                datPayloadLen
-                mnext
-                tlrmkr'
-                sync
-                out
-                reqflush
+            curr datBuf (min datBufSiz lim) >>= \next ->
+              case next of
+                Next datPayloadLen reqflush mnext ->  do
+                  NextTrailersMaker tlrmkr' <- runTrailersMaker tlrmkr datBuf datPayloadLen
+                  fillDataHeaderEnqueueNext
+                      strm
+                      off0
+                      datPayloadLen
+                      mnext
+                      tlrmkr'
+                      sync
+                      out
+                      reqflush
+                CancelNext mErr -> do
+                  -- Stream cancelled
+                  case mErr of
+                    Just err ->
+                      resetStream strm InternalError err
+                    Nothing ->
+                      resetStream strm Cancel (E.toException CancelledStream)
+                  -- It is important that we return the initial offset here,
+                  -- since we might have HEADERS frames in the buffer that
+                  -- /must/ go out before the RST_STREAM
```

Comment:
Although this comment is correct, it's perhaps a bit confusing: the initial offset as opposed to what? If it's "as opposed to 0" (which was our first version here) that's wrong for more reasons than just the headers: it would also affect streams other than the one we are cancelling. 

... Actually, now that I'm trying to formulate a better reason why `off` is right here, I am unsure again about ignoring `_total` (https://github.com/kazu-yamamoto/http-semantics/pull/11#discussion_r1734029953). I need to ponder this a bit more deeply.

Reply:
Ok, I've pondered this more carefully, see https://github.com/kazu-yamamoto/http-semantics/pull/11#issuecomment-2314924295 . I would change the comment above to:

> Return `off0` to discard the current DATA frame under construction. Any other frames constructed previously are left untouched: these may be frames for different streams, or may even be the HEADERS frame for the current stream (which we _must_ send before we can cancel the stream; see `streamHeadersSent`.